### DISCUSSION TITLE
Allow Copy/Pasting and Editing of Filters from the Filter UI.

### DISCUSSION THREAD

**Discussion by Natetronn** [directus/directus#19753] [Feature Requests]
### Summary

Sometimes we write Filters by hand, both for Flows and for filtering the API and sometimes we create them using the Filter GUI to filter Collection items. Why not allow Copy/Pasting and Edit Raw Value of Filters (JSON) from the GUI, this so we don't have to repeat ourselves?

### Basic Example

Just a quick dirty mockup for context [since updated based on conversation thread below]:
 
![filter-menu](https://github.com/directus/directus/assets/582149/f2f4d745-f76d-4388-b73c-119907bfca7a)

### Motivation

Should help speed up the creation of Filters. It's more DRY. Better DX for new user who are learning how Filters work for the first time. Bridging the gap between code and UI. Some people prefer to only code, some prefer to only GUI and others prefer to do both. This feature request is for those who jump back and forth between the GUI and Code.

### Detailed Design

There should be the following options in a drop down menu:

- Copy Raw Value
- Paste Raw Value
- Edit Raw Value
- Undo Changes
- Clear Value

Note: this menu is a clone and inline with the rest of the UI.

It should follow the specs from the Filter-Rules: https://docs.directus.io/reference/filter-rules.html

Paste Raw Value will allow pasting in Filter based JSON and update the UI accordingly. Copy Raw Value will allow copying of JSON based Filters from the current filter in the UI, this JSON can then be use in Flows and or for Filtering the API (rest or sdk). Edit Raw Value will present a window to edit the value directly. Undo and Clear are self-explanatory. 

There should be a safety check to make sure the structure, format and key/values are correct (based on Filter Rules) when Pasting or Editing. If not correct, it should display some sort of error message explaining the root cause of the issue, similar to how editing currently works in code areas.

### Must Have:
- A 
An arrow icon drop down menu, similar to the ones we have for editing field values across Directus. See example mockup above and note the options in detailed design section above.
- B
An error checking system to check Filter based JSON are correctly formatted and importable into the UI. This for Importing. If not, display error message.

### Should Have:
- C
Should have everything needed to make copy, paste, edit, undo and clear filters and should be as seamless as possible for better DX.

### Could Have:
- D

### Won't Have:
- E

### Drawbacks

- implementation cost, both in term of code size and complexity

I can't fully say, but if my assumption is correct, this shouldn't require much effort, since the current filter system (UI) is probably already using something similar under the hood, and the menu we need already exists.

- whether the proposed feature can be implemented in user space

?

- the impact on teaching people Directus

This will lessen the impact on teaching people Directus (Filters)

- integration of this feature with other existing and planned features

?

- cost of migrating existing Directus applications (is it a breaking change?)

I don't believe this is a breaking change (?)

### Alternatives

No other designs have been considered.

### Adoption Strategy

I could be wrong, but I don't believe this will be a breaking change, since it's a UI feature that is built upon an already existing feature set or sets.

No adoption is required, since it's an optional feature.

I don't believe there will need to be a migration (?), but maybe?

Should only affect other projects for the better.

### Unresolved Questions

Would it make sense to present a better working design component in Figma, this to help generate more excitement around such a feature, since there is a [duplicate request](https://github.com/directus/directus/discussions/18913) that got no traction? 

Is the extra effort provided in this FR (vs the old empty one) going to help propel the feature request or will it require more effort on our part?

**Comment by rijkvanzanten** [comment]
I like it! I am wondering if it might be "too technical" for non-technical users 🤔 Maybe this should be an opt-in/opt-out type of thing, or maybe even admin only? Also curious what would happen when you copy a filter and try pasting it on a different collection that may or may not have those same fields.

> Would it make sense to present a better working design component in Figma, this to help generate more excitement around such a feature, since there is a https://github.com/directus/directus/discussions/18913 that got no traction?
> Is the extra effort provided in this FR (vs the old empty one) going to help propel the feature request or will it require more effort on our part?

Out of experience: the more information is available in a feature request, the easier it is for the community to understand and get excited about a feature 🙂

**Reply by Natetronn** [reply]
Thank @rijkvanzanten! 

Yes, maybe it should be an admin only feature?

I'm also wondering if my "Could Have" makes more sense, now that I think about it a bit more. Like, is there really any reason to be importing and exporting actual files? I don't think so. Most of the filters aren't terribly large. I think copy and paste may be the better option, in fact.

It could be a toggle of sorts, where it toggles back and forth between Code Editor JSON View and then of course the current GUI Editor View. Code view could have a JSON/Code editor, with a code copy icon in the corner, like seen in the Directus' docs for code blocks. Devs can just paste in their JSON and select all to copy (or use the copy icon), and be able to edit the Filter code on the fly. Then perhaps it will auto save or there could be a small save button and or when toggling back it saves and updates the GUI View side accordingly etc. 

As a sort of analogy/example, you can see here when making comments on Github they have Write and Preview tabs. You could think of Write being the Code Editor view and Preview being the GUI Editor view (but still editable, like the GUI we have now, of course.) Not saying we'd want to use Tabs, per se, but the toggle between Code/Gui is the point I'm making.

Not to add more to this feature, but I suspect some people would like the GUI to exist in some of the various Flows and Operations. If this was done in such a way, making the whole thing semi portable, perhaps it would make sense for both locations (Flows/Operations filtering and Collection Item filtering). Just a thought, of course, but maybe we could kill two birds with one Filtering system.

Happy to update the visuals with this new idea if it makes more sense to people and isn't terribly that much more difficult to implement?

**Reply by rijkvanzanten** [reply]
I like that idea! Should we extend it beyond filters and have a dedicated button that's more "Show current request as code"? It could be useful to include the actual endpoint with other used query (like sort, pagination) as well 🤔

**Reply by Natetronn** [reply]
I also think it would be very nice to have the end point available somewhere, for quick and easy use in Insomnia or Postman et al. 

That way we can see the Filter's code, have the Filtering URL built out and also edit in cross GUI or Code mode etc. Maybe it could have a "cheatsheet" built into it. They could be buttons based on all the Filter Rules. Clicking or dragging and dropping the buttons could perhaps provide a cross dev/gui filtering builder of sorts. Perhaps this could be a larger modal or drawer that slides out from the right, since it sounds like we'd need more room for all this.

The drawer could be split with a code editor on the left and the "cheatsheet" filter rule "pills" on the right. Or something lol

You could also include some information or links to the various points in the docs on the topic, right in the Filter Builder / Editor.

**Reply by rijkvanzanten** [reply]
This actually ties into another feature request that @benhaynes has been looking to implement for a while now: a proper API explorer in-app. We could treat this as an "inline-preview" of that API explorer, with an expand button to take you to the full "explorer" section

**Reply by Natetronn** [reply]
Kind of reminds me of a Directus focused Nuxt Devtools: 

https://devtools.nuxt.com
https://www.youtube.com/watch?v=Wkla7ATW8Vc

**Reply by benhaynes** [reply]
Yeah, would LOVE to see an API explorer in the App... 3rd party or one we create!

As for this feature, I think it's a bit "power user", but is a nice enhancement. I'd say adding it to a right-click or a "..." context menu would work well though. Not sure if import/export is needed, but Copy/Paste seems nice!

**Reply by Natetronn** [reply]
Yeah, copy/paste seems like the better option, for sure. No need to have extra buttons for export/import, for so little data.

**Reply by Dominic-Marcelino** [reply]
Reminds me of a discussion I opened a while ago: Copying the filters to use them with the sdk 🤔 (https://github.com/directus/directus/discussions/18913)

Apart from that I really like the Idea of having it inside some kind of admin tools like the nuxt devtools

**Reply by Natetronn** [reply]
Hey @Dominic-Marcelino!  Yeah, I found your thread after I had the same idea and came looking for it before posting. But it didn't get any traction at the time and was closed, so I created a new one and added a bit more context in the hopes it would get passed the 15 vote requirement, which it appears to have done, so hopefully we'll see it soon.

Having something like the nuxt devtools is a whole other feature request, but one I'd love to see someday as well.

**Reply by rijkvanzanten** [reply]
Quick unrelated sidebar:

> so I created a new one and added a bit more context in the hopes it would get passed the 15 vote requirement, which it appears to have done

You'd be surprised how much the additional details help people understand and get excited 👀

**Reply by Dominic-Marcelino** [reply]
> But it didn't get any traction at the time and was closed, so I created a new one and added a bit more context in the hopes it would get passed the 15 vote requirement, which it appears to have done, so hopefully we'll see it soon.

More than happy that you did so! Still think it's a super useful feature but at the time i opened the request I didn't had / took the time to outline the advantages as much as you did :)

**Comment by Natetronn** [comment]
Hey everyone, I've updated the Feature Request based on our discussion thread. I think it's a much better solution. Note the mockup, in particular, as it better represents the requirements we discussed.

In its current state, you have a "+ Add Filter", as well as a drop down arrow, which both do the same thing. So I propose we split that in two instead: keep the "+ Add Filter" as is, but for the drop down arrow, it will now reveal a very familiar menu, one that is currently all over Directus i.e. the "Copy/Paste/Edit Value Menu".
